File: py-sdk/src/mcp/protocol/caching.py
"""Response caching hints (§13).

The ``CacheableResult`` fields (``ttlMs`` / ``cacheScope``) and runtime utilities for
validating hints, computing freshness from the client's local clock, resolving the
``cacheScope`` privacy fallback, detecting inconsistent scope across pages, and the
method → invalidation-notification map, plus a minimal in-memory ``ResponseCache``.

Caching hints are purely advisory: they never alter a result's meaning, never function
as access control, and clients MAY ignore them. (R-13.4-f, R-13.3-d, R-13.3-e)
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, Any, Literal

from pydantic import Field, StrictInt

from mcp._model import McpModel, validates

# ─── CacheScope (§13.3) ───────────────────────────────────────────────────────

#: The two sharing-scope values as a field type (the analogue of the TS ``CacheScopeSchema``).
CacheScope = Literal["public", "private"]

#: The two sharing-scope values for a cached result. (§13.3) Unknown/absent → "private".
#:
#: ``"public"`` — any client or shared intermediary MAY reuse the stored copy for any
#: user, subject to the freshness interval. (R-13.3-a)
#:
#: ``"private"`` — may be stored and reused only within the single authorization context
#: that made the request; a shared intermediary MUST NOT serve a stored ``"private"``
#: copy to a different user. (R-13.3-b, R-13.3-c)
#:
#: Any unrecognized or absent value MUST be treated as ``"private"``. (R-13.1-e)
CACHE_SCOPES = ("public", "private")


def is_valid_cache_scope(value: object) -> bool:
  """Return ``True`` only when ``value`` is exactly the case-sensitive string
  ``"public"`` or ``"private"``. (§13.3, R-13.1-d)

  This is the Python analogue of the TS ``CacheScopeSchema`` zod enum: it is strict
  (``"PUBLIC"``, ``"shared"``, ``None``, etc. all fail). Use :func:`resolve_cache_scope`
  when you want the privacy fallback instead of a boolean check.
  """
  return value in CACHE_SCOPES


def _is_int(value: object) -> bool:
  return isinstance(value, int) and not isinstance(value, bool)


def is_cache_hint_valid(ttl_ms: object, cache_scope: object) -> bool:
  """Return ``True`` when BOTH hints are present and valid: non-negative integer
  ``ttlMs`` and ``cacheScope`` exactly ``"public"`` or ``"private"``. (R-13.1-a/-b/-d)
  """
  if not _is_int(ttl_ms) or ttl_ms < 0:
    return False
  return cache_scope in ("public", "private")


def has_both_or_neither_cache_hints(result: dict) -> bool:
  """Return ``True`` when ``result`` carries BOTH hint fields or NEITHER. A server MUST
  NOT emit exactly one without the other. (R-13.1-g)
  """
  return ("ttlMs" in result) == ("cacheScope" in result)


# ─── CacheableResult (§13.1) ──────────────────────────────────────────────────

class CacheableResult(McpModel):
  """A result augmented with the two REQUIRED caching-hint fields (§13.1) — the Python
  analogue of the TS ``CacheableResultSchema``.

  * REQUIRED string ``resultType`` (the §3.6 / S04 base discriminator).
  * REQUIRED non-negative integer ``ttlMs`` (``0`` allowed; negatives / non-integers /
    ``bool`` rejected — ``StrictInt`` accepts only a real ``int``). (R-13.1-a, R-13.2-a)
  * REQUIRED ``cacheScope`` exactly ``"public"`` or ``"private"``. (R-13.1-d)
  * OPTIONAL object ``_meta``.

  Both hint fields MUST be present together (R-13.1-g); method-specific payload members
  (``tools`` / ``contents`` / ``nextCursor`` …) pass through (forward-compatible).
  """

  result_type: str
  ttl_ms: Annotated[StrictInt, Field(ge=0)]
  cache_scope: CacheScope
  meta: dict[str, Any] | None = Field(default=None, alias="_meta")


def is_valid_cacheable_result(value: object) -> bool:
  """Return ``True`` for a valid ``CacheableResult`` wire object. (§13.1)

  The base ``Result`` shape augmented with the two REQUIRED caching-hint fields, both of
  which MUST appear and be valid (see :class:`CacheableResult`). (R-13.1-a/-d/-g)
  """
  return validates(CacheableResult, value)


def resolve_cache_scope(scope: object) -> str:
  """Return ``"public"``/``"private"``, applying the privacy fallback for any
  unrecognized or absent value. (R-13.1-e, R-13.3-h)
  """
  return scope if scope in ("public", "private") else "private"


def is_fresh(ttl_ms: int, received_at: float, now: float) -> bool:
  """Return ``True`` when the result is still within its freshness window:
  ``ttlMs > 0`` AND ``now < received_at + ttlMs``. Uses only the client's local clock.
  (R-13.2-e/-f/-g)
  """
  if ttl_ms <= 0:
    return False
  return now < received_at + ttl_ms


def expires_at(ttl_ms: int, received_at: float) -> float:
  """Return the absolute timestamp after which the result is stale. (R-13.2-e/-f)"""
  return received_at + ttl_ms


def has_consistent_cache_scope(scopes: list[str]) -> bool:
  """Return ``True`` when all ``cacheScope`` values across a list's pages are identical
  (no mixing of public/private). (R-13.5-f/-g)
  """
  if not scopes:
    return True
  return all(s == scopes[0] for s in scopes)


def effective_cache_scope(scopes: list[str]) -> str:
  """Return the scope to apply across a multi-page list; ``"private"`` if inconsistent.
  (R-13.5-h)
  """
  if not has_consistent_cache_scope(scopes):
    return "private"
  # An empty (but trivially consistent) list has no first element; TS reads
  # ``scopes[0]`` as ``undefined`` and resolves to the safe ``"private"`` default.
  return resolve_cache_scope(scopes[0] if scopes else None)


#: Maps each cacheable method to the notification that invalidates its cached result. (§13.5)
METHOD_TO_NOTIFICATION_MAP = {
  "tools/list": "notifications/tools/list_changed",
  "prompts/list": "notifications/prompts/list_changed",
  "resources/list": "notifications/resources/list_changed",
  "resources/templates/list": "notifications/resources/list_changed",
  "resources/read": "notifications/resources/updated",
}


def methods_for_notification(notification: str) -> list[str]:
  """Return the method names whose cached results should be invalidated on
  ``notification``. (R-13.5-j)
  """
  return [m for m, n in METHOD_TO_NOTIFICATION_MAP.items() if n == notification]


@dataclass
class _StoredEntry:
  received_at: float
  ttl_ms: int
  cache_scope: str
  value: dict


@dataclass(frozen=True)
class CacheGetResult:
  """Outcome of :meth:`ResponseCache.get`: ``hit`` plus ``value``/``cache_scope`` on a hit."""

  hit: bool
  value: dict | None = None
  cache_scope: str | None = None


class ResponseCache:
  """A minimal in-memory response cache wired to :func:`is_fresh`,
  :func:`resolve_cache_scope`, and the method→notification invalidation map. (§13)
  """

  def __init__(self) -> None:
    self._store: dict[str, _StoredEntry] = {}

  def set(self, key: str, value: dict, received_at: float) -> None:
    """Store ``value`` under ``key``; skipped when a hint is missing or invalid. A
    ``ttlMs == 0`` entry is stored but never served fresh. (R-13.1-g)
    """
    if not has_both_or_neither_cache_hints(value):
      return
    ttl_ms = value.get("ttlMs")
    raw_scope = value.get("cacheScope")
    if not is_cache_hint_valid(ttl_ms, raw_scope):
      return
    self._store[key] = _StoredEntry(received_at, ttl_ms, resolve_cache_scope(raw_scope), value)

  def get(self, key: str, now: float) -> CacheGetResult:
    """Return the entry for ``key`` if fresh at ``now``; otherwise miss + evict the stale
    entry. (R-13.2-e)
    """
    entry = self._store.get(key)
    if entry is None:
      return CacheGetResult(False)
    if not is_fresh(entry.ttl_ms, entry.received_at, now):
      del self._store[key]
      return CacheGetResult(False)
    return CacheGetResult(True, value=entry.value, cache_scope=entry.cache_scope)

  def invalidate_by_notification(self, notification: str) -> None:
    """Evict all entries (incl. paginated cursor pages) for every method mapped to
    ``notification``. (R-13.5-j)
    """
    for method in methods_for_notification(notification):
      prefix = f"{method}::"
      for key in list(self._store.keys()):
        if key == method or key.startswith(prefix):
          del self._store[key]

  @property
  def size(self) -> int:
    """Number of entries currently stored (may include ``ttlMs == 0`` entries)."""
    return len(self._store)


#: The five methods whose results carry caching hints. (§13.4)
CACHEABLE_METHODS = frozenset(
  {"tools/list", "prompts/list", "resources/list", "resources/templates/list", "resources/read"}
)


def is_cacheable_method(method: str) -> bool:
  """Return ``True`` when ``method`` is one of the five methods that carry caching hints."""
  return method in CACHEABLE_METHODS

File: ts-sdk/src/protocol/caching.ts
/**
 * S19 — Response caching hints (§13).
 *
 * Defines the `CacheableResult` shape and the runtime utilities for:
 *   - Validating caching-hint fields (`ttlMs`, `cacheScope`)
 *   - Computing freshness from the client's local clock
 *   - Resolving `cacheScope` with the privacy fallback
 *   - Detecting inconsistent `cacheScope` across pages of one list
 *   - Identifying the five methods whose results carry hints
 *
 * Caching hints are purely advisory: they never alter a result's meaning,
 * never function as access control, and clients MAY ignore them entirely.
 * (R-13.4-f, R-13.3-d, R-13.3-e)
 */

import { z } from 'zod';
import { ResultTypeSchema } from '../jsonrpc/payload.js';

// ─── CacheScope ───────────────────────────────────────────────────────────────

/**
 * The two sharing-scope values for a cached result. (§13.3)
 *
 * `"public"` — any client or shared intermediary may reuse the stored copy
 * for any user, subject to the freshness interval. (R-13.3-a)
 *
 * `"private"` — may be stored and reused only within the single authorization
 * context that made the request. A shared intermediary MUST NOT serve a stored
 * `"private"` copy to a different user. (R-13.3-b, R-13.3-c)
 *
 * Any unrecognized or absent value MUST be treated as `"private"`. (R-13.1-e)
 */
export const CACHE_SCOPES = ['public', 'private'] as const;
export const CacheScopeSchema = z.enum(CACHE_SCOPES);
export type CacheScope = z.infer<typeof CacheScopeSchema>;

// ─── CacheableResult ─────────────────────────────────────────────────────────

/**
 * A result that augments the base `Result` shape with two advisory caching
 * fields. (§13, R-13.1-a, R-13.1-d)
 *
 * `ttlMs` (REQUIRED on cacheable results): non-negative integer freshness hint
 * in milliseconds. `0` means immediately stale; `N > 0` means fresh for N ms
 * from the client's local receive time. (R-13.2-a, R-13.2-e)
 *
 * `cacheScope` (REQUIRED on cacheable results): `"public"` or `"private"`.
 * (R-13.1-d)
 *
 * Both fields MUST appear together: a server MUST NOT emit one without the
 * other on results specified to carry caching hints. (R-13.1-g)
 *
 * `.passthrough()` preserves method-specific payload members.
 */
export const CacheableResultSchema = z
  .object({
    /** REQUIRED base discriminator. (§3.6 / S04) */
    resultType: ResultTypeSchema,
    /** OPTIONAL metadata. */
    _meta: z.record(z.unknown()).optional(),
    /**
     * REQUIRED non-negative integer freshness hint in milliseconds. (R-13.1-a, R-13.2-a)
     * `0` = immediately stale; `N > 0` = fresh for N ms from local receive time.
     */
    ttlMs: z.number().int().nonnegative(),
    /**
     * REQUIRED sharing scope. Must be exactly `"public"` or `"private"`.
     * (R-13.1-d)
     */
    cacheScope: CacheScopeSchema,
  })
  .passthrough();

export type CacheableResult = z.infer<typeof CacheableResultSchema>;

// ─── Hint validation ──────────────────────────────────────────────────────────

/**
 * Returns `true` when BOTH caching hint fields are present and valid.
 * (R-13.1-a, R-13.1-b, R-13.1-d)
 *
 * A receiver MUST NOT treat a result as cacheable when `ttlMs` is negative,
 * non-integer, or missing, and MUST treat `cacheScope` as `"private"` when
 * the value is unrecognized or missing. (R-13.1-b, R-13.1-e)
 *
 * @param ttlMs - The raw value of the `ttlMs` field from the result.
 * @param cacheScope - The raw value of the `cacheScope` field from the result.
 */
export function isCacheHintValid(ttlMs: unknown, cacheScope: unknown): boolean {
  if (
    typeof ttlMs !== 'number' ||
    !Number.isInteger(ttlMs) ||
    ttlMs < 0
  ) {
    return false;
  }
  return cacheScope === 'public' || cacheScope === 'private';
}

/**
 * Returns `true` when a result object carries BOTH caching-hint fields (or
 * neither). A server MUST NOT emit exactly one without the other. (R-13.1-g)
 *
 * Pass a raw result object; this is a conformance check on server output.
 */
export function hasBothOrNeitherCacheHints(result: Record<string, unknown>): boolean {
  const hasTtl = 'ttlMs' in result;
  const hasScope = 'cacheScope' in result;
  return hasTtl === hasScope;
}

// ─── CacheScope resolution ────────────────────────────────────────────────────

/**
 * Returns `"public"` or `"private"`, applying the privacy fallback for any
 * unrecognized or absent value. (R-13.1-e, R-13.3-h)
 *
 * A receiver that cannot reliably distinguish authorization contexts MUST treat
 * every cached result as `"private"`.
 */
export function resolveCacheScope(scope: unknown): CacheScope {
  if (scope === 'public' || scope === 'private') return scope;
  return 'private';
}

// ─── Freshness computation ────────────────────────────────────────────────────

/**
 * Returns `true` when the result is still within its freshness window.
 * (R-13.2-e, R-13.2-f)
 *
 * Formula: `(ttlMs > 0) AND (now < receivedAt + ttlMs)`.
 *
 * A client MUST NOT assume the client and server clocks agree; the computation
 * uses only the client's local `receivedAt` and the `ttlMs` value. (R-13.2-g)
 *
 * @param ttlMs - Non-negative freshness hint from the result.
 * @param receivedAt - The client's local timestamp (ms since epoch) when the
 *   response was received.
 * @param now - The current client-local timestamp (ms since epoch).
 */
export function isFresh(ttlMs: number, receivedAt: number, now: number): boolean {
  if (ttlMs <= 0) return false;
  return now < receivedAt + ttlMs;
}

/**
 * Computes `expiresAt` — the absolute timestamp after which the result is stale.
 * (R-13.2-e, R-13.2-f)
 *
 * @param ttlMs - Non-negative freshness hint.
 * @param receivedAt - Local receive timestamp in ms.
 */
export function expiresAt(ttlMs: number, receivedAt: number): number {
  return receivedAt + ttlMs;
}

// ─── Page-scope consistency ───────────────────────────────────────────────────

/**
 * Returns `true` when all `cacheScope` values across the pages of one logical
 * list are identical (no mixing of `"public"` and `"private"`). (R-13.5-f, R-13.5-g)
 *
 * A server MUST NOT mix `"public"` and `"private"` across pages. When a client
 * observes inconsistency it MUST treat the entire list as `"private"`. (R-13.5-h)
 */
export function hasConsistentCacheScope(scopes: ReadonlyArray<string>): boolean {
  if (scopes.length === 0) return true;
  const first = scopes[0];
  return scopes.every((s) => s === first);
}

/**
 * Given the `cacheScope` values observed across a multi-page list, returns the
 * effective scope to apply. If inconsistent, returns `"private"`. (R-13.5-h)
 */
export function effectiveCacheScope(scopes: ReadonlyArray<string>): CacheScope {
  if (!hasConsistentCacheScope(scopes)) return 'private';
  return resolveCacheScope(scopes[0]);
}

// ─── Method → notification invalidation map ───────────────────────────────────

/**
 * Maps each cacheable method name to the notification that signals a change.
 * When the notification arrives the client MUST discard the cached result and
 * re-fetch. (§13.5, R-13.5-a)
 */
export const METHOD_TO_NOTIFICATION_MAP: Readonly<Record<string, string>> = {
  'tools/list': 'notifications/tools/list_changed',
  'prompts/list': 'notifications/prompts/list_changed',
  'resources/list': 'notifications/resources/list_changed',
  'resources/templates/list': 'notifications/resources/list_changed',
  'resources/read': 'notifications/resources/updated',
};

/**
 * Returns the method names whose cached results should be invalidated when
 * `notification` is received. (§13.5, R-13.5-a)
 */
export function methodsForNotification(notification: string): ReadonlyArray<string> {
  return Object.entries(METHOD_TO_NOTIFICATION_MAP)
    .filter(([, n]) => n === notification)
    .map(([m]) => m);
}

// ─── ResponseCache ────────────────────────────────────────────────────────────

interface StoredEntry {
  readonly receivedAt: number;
  readonly ttlMs: number;
  readonly cacheScope: CacheScope;
  readonly value: Record<string, unknown>;
}

/** Result type returned by `ResponseCache.get`. */
export type CacheGetResult<T> =
  | { hit: true; value: T; cacheScope: CacheScope }
  | { hit: false };

/**
 * Minimal in-memory response cache wired to `isFresh`, `resolveCacheScope`,
 * and the method→notification invalidation map. (§13, R-13.5-j)
 *
 * - Freshness is computed via `isFresh(ttlMs, receivedAt, now)` — `ttlMs=0`
 *   entries are stored but never served fresh. (RC-3)
 * - `invalidateByNotification` evicts all entries (including paginated cursor
 *   pages) for every method mapped to the given notification. (RC-9)
 * - Scope is resolved conservatively via `resolveCacheScope`. (RC-5)
 */
export class ResponseCache<T extends Record<string, unknown>> {
  private readonly store = new Map<string, StoredEntry>();

  /**
   * Stores `value` under `key`. Skipped when either caching hint is missing or
   * invalid (`hasBothOrNeitherCacheHints` + `isCacheHintValid`). A `ttlMs=0`
   * entry is stored but will never be returned as a cache hit. (R-13.1-g)
   */
  set(key: string, value: T, receivedAt: number): void {
    if (!hasBothOrNeitherCacheHints(value as Record<string, unknown>)) return;
    const ttlMs = value['ttlMs'];
    const rawScope = value['cacheScope'];
    if (!isCacheHintValid(ttlMs, rawScope)) return;
    this.store.set(key, {
      value: value as Record<string, unknown>,
      receivedAt,
      ttlMs: ttlMs as number,
      cacheScope: resolveCacheScope(rawScope),
    });
  }

  /**
   * Returns the entry for `key` if it is still fresh at `now`; otherwise returns
   * `{ hit: false }` and evicts the stale entry. (R-13.2-e, RC-3, RC-5)
   */
  get(key: string, now: number): CacheGetResult<T> {
    const entry = this.store.get(key);
    if (!entry) return { hit: false };
    if (!isFresh(entry.ttlMs, entry.receivedAt, now)) {
      this.store.delete(key);
      return { hit: false };
    }
    return { hit: true, value: entry.value as T, cacheScope: entry.cacheScope };
  }

  /**
   * Evicts all entries for every method that maps to `notification`, including
   * all paginated-cursor-page entries (keys prefixed with `method::`). (RC-9)
   */
  invalidateByNotification(notification: string): void {
    const methods = methodsForNotification(notification);
    for (const method of methods) {
      const prefix = `${method}::`;
      for (const key of [...this.store.keys()]) {
        if (key === method || key.startsWith(prefix)) {
          this.store.delete(key);
        }
      }
    }
  }

  /** Number of entries currently stored (may include `ttlMs=0` entries). */
  get size(): number {
    return this.store.size;
  }
}

// ─── Cacheable methods registry ───────────────────────────────────────────────

/**
 * The set of method names whose results carry `CacheableResult` shapes. (§13.4, R-13.4-a)
 *
 * On every result from these methods a server MUST populate both `ttlMs` and
 * `cacheScope` with valid values. A server that does not wish to encourage
 * caching MUST still include the fields and SHOULD set `ttlMs` to `0`. (R-13.4-b)
 *
 * On any other message, receivers MUST ignore `ttlMs`/`cacheScope` if present.
 * (R-13.4-e)
 */
export const CACHEABLE_METHODS = new Set([
  'tools/list',
  'prompts/list',
  'resources/list',
  'resources/templates/list',
  'resources/read',
] as const);

/** Returns `true` when `method` is one of the five methods that carry caching hints. */
export function isCacheableMethod(method: string): boolean {
  return CACHEABLE_METHODS.has(method as never);
}
